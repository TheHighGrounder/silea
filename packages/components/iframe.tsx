/* eslint-disable jsx-a11y/iframe-has-title */
import React from "react";
import { connect, Head } from "silea";
import { Connect, Package } from "silea/types";
import useInView from "@sileajs/hooks/use-in-view";

const noJsStyles = `
  :non(noscript) > .silea-lazy-iframe {
    display: none;
  }
`;

const noProxyScript = `
  if (typeof window !== "undefined" && (!("Proxy" in window) || !("IntersectionObserver" in window))) {
    document.addEventListener("DOMContentLoaded", function() {
      var iframes = document.querySelectorAll("iframe.silea-lazy-iframe");
      for (i = 0; i < iframes.length; ++i) {
        var iframe = iframes[i];
        iframe.setAttribute("src", iframe.getAttribute("data-src"));
        iframe.removeAttribute("data-src");
        iframe.removeAttribute("style");
      }
    });
  }
`;

interface Props {
	title: string;
	src?: string;
	width?: number;
	height?: number;
	className?: string;
	rootMargin?: string;
	loading?: "lazy" | "eager";
}

type Component = React.FC<Connect<Package, Props>>;

interface Attributes extends Props {
	"data-src"?: string;
	style?: { visibility: "hidden" };
}

type NoScriptIframe = React.FC<Attributes>;

interface ChangeAttributes {
	(attrs: Attributes): Attributes;
}

const changeAttributes: ChangeAttributes = (attrs) => {
	const attributes = { ...attrs };

	attributes.src = attributes["data-src"];

	delete attributes["data-src"];
	delete attributes["style"];

	return attributes;
};

const NoScriptIframe: NoScriptIframe = (props) => {
	const attributes = { ...props };

	return (
		<noscript>
			<iframe {...attributes} />
		</noscript>
	);
};

const Iframe: Component = ({
	state,
	src,
	title,
	width,
	height,
	className,
	loading = "lazy",
	rootMargin,
}) => {
	const lazyAttributes: Attributes = {
		"data-src": src,
		className: "silea-lazy-iframe".concat(className ? ` ${className}` : ""),
		loading,
		style: { visibility: "hidden" },
		height,
		width,
		title,
	};

	const eagerAttributes = changeAttributes(lazyAttributes);

	const { ref, inView, supported } = useInView({
		rootMargin,
		triggerOnce: true,
	});

	if (loading === "eager") return <iframe {...eagerAttributes} />;

	if (typeof window !== "undefined") {
		if (
			supported &&
			!("loading" in HTMLIFrameElement.prototype && height > 0)
		) {
			return (
				<>
					<NoScriptIframe {...eagerAttributes} />
					<iframe
						ref={ref}
						{...(inView ? eagerAttributes : lazyAttributes)}
					/>
				</>
			);
		}

		return (
			<>
				<NoScriptIframe {...eagerAttributes} />
				<iframe
					{...(state.silea.rendering === "csr"
						? eagerAttributes
						: lazyAttributes)}
				/>
			</>
		);
	}

	return (
		<>
			<Head
				script={[
					{
						id: "silea-no-proxy-iframe",
						type: "text/javascript",
						innerHTML: noProxyScript,
					},
				]}
				noscript={[
					{
						innerHTML: `<style id="silea-no-js-iframes" type="text/css">${noJsStyles}</style>`,
					},
				]}
			/>
			<NoScriptIframe {...eagerAttributes} />
			<iframe {...lazyAttributes} />
		</>
	);
};

export default connect(Iframe);
