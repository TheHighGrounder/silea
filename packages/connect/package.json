{
	"name": "@sileajs/connect",
	"version": "0.0.2",
	"description": "The state manager of Silea. Similar to MobX but much more lightweight because it's based on ES6 proxies.",
	"keywords": [
		"silea",
		"connect",
		"state-manager"
	],
	"homepage": "https://sileajs.com",
	"license": "SEE LICENSE IN LICENSE",
	"main": "src/index.js",
	"types": "src/index.d.ts",
	"repository": {
		"type": "git",
		"url": "https://github.com/SileaJS/silea.git"
	},
	"scripts": {
		"test:ci": "../../node_modules/.bin/jest --ci --coverage",
		"test": "../../node_modules/.bin/jest --watch",
		"test:inspect": "node --inspect ../../node_modules/.bin/jest --watch --no-cache --runInBand"
	},
	"dependencies": {
		"@sileajs/error": "0.0.2"
	},
	"devDependencies": {
		"react-test-renderer": "16.13.1"
	},
	"publishConfig": {
		"access": "public"
	}
}
