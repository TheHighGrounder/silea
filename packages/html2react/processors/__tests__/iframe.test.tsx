import React from "react";
import { render } from "@testing-library/react";
import { createStore } from "@sileajs/connect";
import { Html2React } from "../../src/libraries/component";
import iframe from "../iframe";
import Html2ReactPackage from "../../src";

const store = createStore(Html2ReactPackage);
store.libraries.html2react.processors.push(iframe);

jest.mock("@sileajs/components/iframe", () => "mocked-iframe");

describe("Iframe Processor", () => {
	it("should process iframes from src", () => {
		const { container } = render(
			<Html2React
				html="<iframe src='https://sileajs.com' title='Silea' width='500' />"
				{...store}
			/>
		);

		expect(container).toMatchSnapshot();
	});

	it("should process iframes from data-src", () => {
		const { container } = render(
			<Html2React
				html="<iframe data-src='https://sileajs.com' title='Silea' width='500' />"
				{...store}
			/>
		);

		expect(container).toMatchSnapshot();
	});
});
