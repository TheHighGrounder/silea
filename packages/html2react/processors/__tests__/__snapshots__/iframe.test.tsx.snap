// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Iframe Processor should process iframes from data-src 1`] = `
<div>
  <mocked-iframe
    data-src="https://sileajs.com"
    src="https://sileajs.com"
    title="Silea"
    width="500"
  />
</div>
`;

exports[`Iframe Processor should process iframes from src 1`] = `
<div>
  <mocked-iframe
    src="https://sileajs.com"
    title="Silea"
    width="500"
  />
</div>
`;
